File: android/app/src/main/java/com/bkes994408/expensetracker/pro/BudgetModels.kt
package com.bkes994408.expensetracker.pro

import kotlin.math.max

enum class CarryOverMode {
    NONE,
    ROLLOVER,
}

data class BudgetPlan(
    val id: Long,
    val monthKey: String,
    val categoryName: String,
    val amount: Double,
    val carryOverMode: CarryOverMode,
)

enum class BudgetStatus {
    HEALTHY,
    WARNING,
    OVERSPENT,
}

data class BudgetProgress(
    val categoryName: String,
    val spent: Double,
    val budget: Double,
) {
    val remaining: Double = budget - spent
    val ratio: Double = if (budget <= 0.0) 0.0 else max(0.0, spent / budget)
    val status: BudgetStatus = when {
        ratio > 1.0 -> BudgetStatus.OVERSPENT
        ratio >= 0.8 -> BudgetStatus.WARNING
        else -> BudgetStatus.HEALTHY
    }
}

File: ios/ExpenseTracker/Budget/BudgetModels.swift
import Foundation

enum CarryOverMode: String, CaseIterable {
    case none
    case rollover
}

struct BudgetPlan: Identifiable, Equatable {
    let id: Int64
    let monthKey: String
    let categoryName: String
    let amount: Decimal
    let carryOverMode: CarryOverMode
}

struct BudgetProgress: Identifiable, Equatable {
    enum Status: Equatable {
        case healthy
        case warning
        case overspent
    }

    let id: Int64
    let categoryName: String
    let spent: Decimal
    let budget: Decimal

    var remaining: Decimal { budget - spent }
    var ratio: Double {
        guard budget > 0 else { return 0 }
        let spentDouble = NSDecimalNumber(decimal: spent).doubleValue
        let budgetDouble = NSDecimalNumber(decimal: budget).doubleValue
        return max(0, spentDouble / budgetDouble)
    }

    var status: Status {
        if ratio > 1 { return .overspent }
        if ratio >= 0.8 { return .warning }
        return .healthy
    }
}
